File: apps/api/app/core/sentry.py
"""
Sentry integration for error tracking and monitoring
"""

import sentry_sdk
from sentry_sdk.integrations.fastapi import FastApiIntegration
from sentry_sdk.integrations.sqlalchemy import SqlalchemyIntegration
from sentry_sdk.integrations.logging import LoggingIntegration

from app.core.config import settings
from app.core.logging import get_logger

logger = get_logger(__name__)


def init_sentry():
    """
    Initialize Sentry SDK
    
    Features:
    - Error tracking
    - Performance monitoring
    - Request correlation
    - SQL query tracking
    - User context
    """
    if not settings.SENTRY_DSN:
        logger.info("Sentry DSN not configured - error tracking disabled")
        return
    
    sentry_sdk.init(
        dsn=settings.SENTRY_DSN,
        
        # Environment
        environment=settings.ENVIRONMENT,
        release=f"{settings.APP_NAME}@{settings.APP_VERSION}",
        
        # Performance monitoring
        traces_sample_rate=1.0 if settings.is_development else 0.1,
        profiles_sample_rate=1.0 if settings.is_development else 0.1,
        
        # Integrations
        integrations=[
            FastApiIntegration(
                transaction_style="endpoint",  # Group by endpoint
            ),
            SqlalchemyIntegration(),
            LoggingIntegration(
                level=None,  # Capture all logs
                event_level="ERROR",  # Send ERROR+ to Sentry
            ),
        ],
        
        # Additional options
        attach_stacktrace=True,
        send_default_pii=False,  # Don't send PII automatically
        max_breadcrumbs=50,
        
        # Filter sensitive data
        before_send=filter_sensitive_data,
    )
    
    logger.info(f"Sentry initialized: {settings.ENVIRONMENT}")


def filter_sensitive_data(event, hint):
    """
    Filter sensitive data before sending to Sentry
    
    Removes:
    - Authorization headers
    - Password fields
    - API keys
    - Personal information
    """
    # Remove sensitive headers
    if "request" in event and "headers" in event["request"]:
        sensitive_headers = ["authorization", "cookie", "x-api-key"]
        headers = event["request"]["headers"]
        
        for header in sensitive_headers:
            if header in headers:
                headers[header] = "[Filtered]"
    
    # Remove sensitive query parameters
    if "request" in event and "query_string" in event["request"]:
        query = event["request"]["query_string"]
        if any(key in query.lower() for key in ["token", "key", "password"]):
            event["request"]["query_string"] = "[Filtered]"
    
    # Remove sensitive body fields
    if "request" in event and "data" in event["request"]:
        sensitive_fields = ["password", "password_hash", "api_key", "secret"]
        data = event["request"]["data"]
        
        if isinstance(data, dict):
            for field in sensitive_fields:
                if field in data:
                    data[field] = "[Filtered]"
    
    return event


def set_user_context(user_id: str, email: str = None, username: str = None):
    """
    Set user context for Sentry events
    
    Args:
        user_id: User ID
        email: User email (optional)
        username: Username (optional)
    """
    sentry_sdk.set_user({
        "id": user_id,
        "email": email,
        "username": username,
    })


def add_breadcrumb(message: str, category: str = "info", level: str = "info", data: dict = None):
    """
    Add breadcrumb for debugging
    
    Breadcrumbs help trace the path leading to an error
    """
    sentry_sdk.add_breadcrumb(
        message=message,
        category=category,
        level=level,
        data=data or {},
    )


def capture_exception(exception: Exception, context: dict = None):
    """
    Manually capture exception
    
    Args:
        exception: Exception to capture
        context: Additional context data
    """
    if context:
        sentry_sdk.set_context("custom", context)
    
    sentry_sdk.capture_exception(exception)


def capture_message(message: str, level: str = "info", context: dict = None):
    """
    Capture message (not an exception)
    
    Useful for logging important events
    """
    if context:
        sentry_sdk.set_context("custom", context)
    
    sentry_sdk.capture_message(message, level)

File: apps/miniapp/src/lib/sentry.ts
/**
 * Sentry integration for frontend error tracking
 */

import * as Sentry from "@sentry/react";
import { BrowserTracing } from "@sentry/tracing";

const SENTRY_DSN = import.meta.env.VITE_SENTRY_DSN;
const ENVIRONMENT = import.meta.env.VITE_ENVIRONMENT || "development";

export function initSentry() {
  if (!SENTRY_DSN) {
    console.log("Sentry DSN not configured - error tracking disabled");
    return;
  }

  Sentry.init({
    dsn: SENTRY_DSN,
    
    // Environment
    environment: ENVIRONMENT,
    
    // Performance monitoring
    integrations: [
      new BrowserTracing(),
      new Sentry.Replay({
        maskAllText: true,
        blockAllMedia: true,
      }),
    ],
    
    // Sample rates
    tracesSampleRate: ENVIRONMENT === "production" ? 0.1 : 1.0,
    replaysSessionSampleRate: 0.1,
    replaysOnErrorSampleRate: 1.0,
    
    // Filter errors
    beforeSend(event, hint) {
      // Filter out known non-critical errors
      const error = hint.originalException;
      
      if (error && typeof error === "object") {
        const message = error.toString();
        
        // Ignore network errors (handled by UI)
        if (message.includes("NetworkError") || message.includes("Failed to fetch")) {
          return null;
        }
        
        // Ignore ResizeObserver errors (browser bug)
        if (message.includes("ResizeObserver")) {
          return null;
        }
      }
      
      // Remove sensitive data
      if (event.request?.headers) {
        delete event.request.headers["Authorization"];
        delete event.request.headers["Cookie"];
      }
      
      return event;
    },
  });

  console.log("Sentry initialized:", ENVIRONMENT);
}

export function setUser(userId: string, email?: string) {
  Sentry.setUser({
    id: userId,
    email: email,
  });
}

export function clearUser() {
  Sentry.setUser(null);
}

export function captureException(error: Error, context?: Record<string, any>) {
  if (context) {
    Sentry.setContext("custom", context);
  }
  Sentry.captureException(error);
}

export function captureMessage(message: string, level: Sentry.SeverityLevel = "info") {
  Sentry.captureMessage(message, level);
}

export function addBreadcrumb(message: string, category: string = "user-action") {
  Sentry.addBreadcrumb({
    message,
    category,
    level: "info",
  });
}

